src/data/ordmap/ordnode.lean

```diff
@@ -0,0 +1,1152 @@
+    { refine if rs > delta * ls then _ else node (ls + rs + 1) l x r,
+      cases id rr with rrs, {exact nil /-should not happen-/},
+      cases id rl with rls rll rlx rlr, {exact nil /-should not happen-/},
+      exact if rls < ratio * rrs then
+        node (ls + rs + 1) (node (ls + rls + 1) l x rl) rx rr
+      else
+        node (ls + rs + 1)
+          (node (ls + size rll + 1) l x rll) rlx
+          (node (size rlr + rrs + 1) rlr rx rr) } }
+end
+
+/-- O(1). Rebalance a tree which was previously balanced but has had one side change
+by at most 1. Internal use only. -/
+def balance (l : ordnode α) (x : α) (r : ordnode α) : ordnode α :=
+by clean begin
+  cases id l with ls ll lx lr,
+  { cases id r with rs rl rx rr,
+    { exact ι x },
+    { cases id rl with rls rll rlx rlr,
+      { cases id rr,
+        { exact node 2 nil x r },
+        { exact node 3 (ι x) rx rr } },
+      { cases id rr with rrs,
+        { exact node 3 (ι x) rlx (ι rx) },
+        { exact if rls < ratio * rrs then
+            node (rs+1) (node (rls+1) nil x rl) rx rr
+          else
+            node (rs+1)
+              (node (size rll + 1) nil x rll) rlx
+              (node (size rlr + rrs + 1) rlr rx rr) } } } },
+  { cases id r with rs rl rx rr,
+    { cases id ll with lls,
+      { cases lr with _ _ lrx,
+        { exact node 2 l x nil },
+        { exact node 3 (ι lx) lrx (ι x) } },
+      { cases id lr with lrs lrl lrx lrr,
+        { exact node 3 ll lx (ι x) },
+        { exact if lrs < ratio * lls then
+            node (ls+1) ll lx (node (lrs+1) lr x nil)
+          else
+            node (ls+1)
+              (node (lls + size lrl + 1) ll lx lrl) lrx
+              (node (size lrr + 1) lrr x nil) } } },
+    { refine
+        if delta * ls < rs then _ else
+        if delta * rs < ls then _ else
+        node (ls+rs+1) l x r,
+      { cases id rl with rls rll rlx rlr, {exact nil /-should not happen-/},
+        cases id rr with rrs, {exact nil /-should not happen-/},
+        exact if rls < ratio * rrs then
+          node (ls+rs+1) (node (ls+rls+1) l x rl) rx rr
+        else
+          node (ls+rs+1)
+            (node (ls + size rll + 1) l x rll) rlx
+            (node (size rlr + rrs + 1) rlr rx rr) },
+      { cases id ll with lls, {exact nil /-should not happen-/},
+        cases id lr with lrs lrl lrx lrr, {exact nil /-should not happen-/},
+        exact if lrs < ratio * lls then
+          node (ls+rs+1) ll lx (node (lrs+rs+1) lr x r)
+        else
+          node (ls+rs+1)
+            (node (lls + size lrl + 1) ll lx lrl) lrx
+            (node (size lrr + rs + 1) lrr x r) } } }
+end
+
+/-- O(n). Does every element of the map satisfy property `P`?
+
+     all (λ x, x < 5) {1, 2, 3} = true
+     all (λ x, x < 5) {1, 2, 3, 5} = false -/
+def all (P : α → Prop) : ordnode α → Prop
+| nil := true
+| (node _ l x r) := all l ∧ P x ∧ all r
+
+instance all.decidable {P : α → Prop} [decidable_pred P] (t) : decidable (all P t) :=
```

Comment:
I guess this and below are special cases of a derive handler for decidable predicates, but that's future work.